File: Algo/Leetcode/049GroupAnagram.cpp
// Use a dict with key-vec(string)
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        if(strs.size()==0) return res;
        map<string, vector<string>> m;
        for(auto s:strs){
            string k(s.begin(), s.end());
            sort(k.begin(), k.end());
            if(m.find(k) == m.end()) m[k] = vector<string>{s};
            else m[k].push_back(s);
        }
        for(auto it=m.begin(); it!=m.end(); it++){
            res.push_back(it->second);
        }
        return res;
    }
};

File: Algo/Leetcode/049GroupAnagram.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        d = {}
        for s in strs:
            s_sort = ''.join(sorted(s))     # sorted string is list
            if d.get(s_sort) is not None:
                d[s_sort].append(s)
            else:
                d[s_sort] = [s]

        res = []
        for v in d.values():
            res.append(v)

        return res